### PR TITLE
Bug 1409090 - Long-press reload button to disable content blocker

### PR DESCRIPTION
**Pull request by garvankeeley**
https://bugzilla.mozilla.org/show_bug.cgi?id=1409090

### CODE REVIEW

**Comment by garvankeeley** [comment]
Screenshots on iPhone 5:
Long press to show option to disable
simulator screen shot - iphone 5s - 2017-11-08 at 16 51 15

Followed by another long press to enable. If you look behind the cancel button a FedEx ad is showing:
simulator screen shot - iphone 5s - 2017-11-08 at 16 51 34

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift [hunk 1]
```diff
@@ -20,6 +20,24 @@ class ContentBlockerHelper {
...
+        case disallowUserOverride // Option is not offered if tracking protection is off in prefs
+        case allowedButNotSet
+        case forceEnabled
+        case forceDisabled
+    }
+
+    fileprivate var prefOverrideTrackingProtectionEnabled: Bool?
+    var userOverrideForTrackingProtection: TrackingProtectionUserOverride {
+        if !trackingProtectionEnabledInSettings {
+            return .disallowUserOverride
```
Comment (garvankeeley) [inline Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: I like the idea of having long-press to "Reload with TP" even if TP is off in the app settings.
For this patch I didn't do that. I am filing a follow up bug to propose that.

**Comment by garvankeeley** [comment]
Need auto screenshots for this!

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift (hunk 1 above)
Comment (jhugman) [reply Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: This is my only feedback for this PR.

Comment (garvankeeley) [reply Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: thanks, filed for follow up https://bugzilla.mozilla.org/show_bug.cgi?id=1409090#c10

**Comment by garvankeeley** [comment]
I changed the screenshots tests to do:
```
    func test13ReloadButtonContextMenu() {
        // TP on for first screenshots
        navigator.goto(TrackingProtectionSettings)
        app.tables.staticTexts["Always On"].tap()
        navigator.goto(BrowserTab)

        navigator.openURL(loremIpsumURL)
        navigator.toggleOff(userState.requestDesktopSite, withAction: Action.ToggleRequestDesktopSite)
        navigator.goto(ReloadLongPressMenu)
        snapshot("13ContextMenuReloadButton-01")
        navigator.toggleOn(userState.requestDesktopSite, withAction: Action.ToggleRequestDesktopSite)
        navigator.goto(ReloadLongPressMenu)
        snapshot("13ContextMenuReloadButton-02", waitForLoadingIndicator: false)

        // TP off for screenshot
        app.sheets.buttons["Reload Without Tracking Protection"].tap()
        navigator.goto(ReloadLongPressMenu) // HELP! this doesn't re-show the menu
        snapshot("13ContextMenuReloadButton-03", waitForLoadingIndicator: false)
    }
```

**Comment by jhugman** [comment]
Look how it's done with `requestDesktopSite`. There's a `Bool` in `FxUserState`, and an action in the `ReloadLongPressMenu`. 

*Bonus*: IIRC, toggling tracking protection isn't available if tracking protection isn't on from the settings, so you may want to add an `if:` argument for the `Action.toggleTrackingProtection`.

The simplest explanation of what is going on:

```swift
        // navigator is now at BrowserTab.
        navigator.goto(ReloadLongPressMenu) // now at ReloadLongPressMenu.
        snapshot("13ContextMenuReloadButton-02", waitForLoadingIndicator: false)

        // TP off for screenshot
        app.sheets.buttons["Reload Without Tracking Protection"].tap() // navigator hasn't moved, so thinks we're still at ReloadLongPressMenu.
        navigator.goto(ReloadLongPressMenu) // Nothing happens, we're already there.
```

The simplest work around is to tell the navigator: `navigator.nowAt(BrowserTab)`, but really you should add:

```swift

    map.addScreenState(ReloadLongPressMenu) { screenState in
        …
        let trackingProtectionButton = app.sheets.buttons["Reload Without Tracking Protection"] // use accessibilityIdentifier please.
        screenState.tap(trackingProtectionButton, forAction: Action.ToggleTrackingProtection) { userState in
            userState.trackingProtection = !userState.trackingProtection
        }
        …
    }
```

with corresponding flags in `FxUserState`.

**Comment by garvankeeley** [comment]
Landing, tests will follow in https://bugzilla.mozilla.org/show_bug.cgi?id=1417659
The testing patch is turning out to be substantial.

**Pull request merged**